![1001 genomes](http://1001genomes.org/layout_files/logo_1001genomes_v5.png)

# *Arabidopsis thaliana* 1001 Genomes Project

<!-- TOC -->
* [*Arabidopsis thaliana* 1001 Genomes Project](#arabidopsis-thaliana-1001-genomes-project)
  * [Basic info](#basic-info)
  * [Project](#project)
  * [Download](#download)
    * [Reference](#reference)
    * [Illumina](#illumina)
  * [Symlink](#symlink)
  * [Run](#run)
  * [Pack and clean](#pack-and-clean)
  * [VCF](#vcf)
<!-- TOC -->

## Basic info

* Genome: GCF_000001735.3, TAIR10, 119.668 Mb
* Chloroplast: [NC_000932](https://www.ncbi.nlm.nih.gov/nuccore/NC_000932), **Columbia**, 154478 bp
* Chloroplast: [KX551970](https://www.ncbi.nlm.nih.gov/nuccore/KX551970), **Landsberg erecta**,
  154515 bp
* Mitochondrion: [Y08501](https://www.ncbi.nlm.nih.gov/nuccore/Y08501), 366924 bp

## Project

* PRJNA273563

> The 1001 Genomes Project was launched at the beginning of 2008 to discover the whole-genome
> sequence variation in 1001 accessions of the reference plant Arabidopsis thaliana. The resulting
> information is paving the way for a new era of genetics that identifies alleles underpinning
> phenotypic diversity across the entire genome and the entire species. Each of the accessions in
> the 1001 Genomes project is an inbred line with seeds that are freely available from the stock
> centre to all our colleagues. Unlimited numbers of plants with identical genotype can be grown and
> phenotyped for each accession, in as many environments as desired, and so the sequence information
> we collect can be used directly in association studies at biochemical, metabolic, physiological,
> morphological, and whole plant-fitness levels. The analyses enabled by this project will have
> broad implications for areas as diverse as evolutionary sciences, plant breeding and human
> genetics.
>
> 1001 基因组计划于 2008 年初启动, 旨在发现参考植物拟南芥中1001个基因组序列的变异.
> 由此产生的信息正在为一个新的遗传学时代铺平道路, 该时代将识别支撑整个基因组和整个物种表型多样性的等位基因.
> 1001 基因组计划的每一项研究都是一种自交系, 种子可以从库存中心免费获得给我们所有的同事.
> 无限数量的具有相同基因型的植物可以在任意多的环境中生长和表型,
> 因此我们收集的序列信息可以直接用于生物化学、代谢、生理、形态学和整个植物适应性水平的关联研究.
> 该项目的分析将对进化科学、植物育种和人类遗传学等不同领域产生广泛的影响.

## Download

### Reference

```shell
mkdir -p ~/data/plastid/Atha_1001/genome
cd ~/data/plastid/Atha_1001/genome

for ACCESSION in "NC_000932" "Y08501"; do
    URL=$(printf "https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=nucleotide&rettype=%s&id=%s&retmode=text" "fasta" "${ACCESSION}");
    curl $URL -o ${ACCESSION}.fa
done

TAB=$'\t'
cat <<EOF > replace.tsv
NC_000932${TAB}Pt
Y08501${TAB}Mt
EOF

cat NC_000932.fa Y08501.fa |
    hnsm filter -s stdin |
    hnsm replace stdin replace.tsv |
    hnsm order stdin <(echo Pt; echo Mt) -o genome.fa

```

### Illumina

* Download `Metadata` from NCBI SRA Run Selector via a web browser
    * <https://trace.ncbi.nlm.nih.gov/Traces/study/?acc=PRJNA273563>
    * Save it to `SraRunTable.csv`

* Download Accessions info
    * <http://1001genomes.org/accessions.html>
    * There's no header line
    * Rename `Accession ID` to `Ecotype`

```shell
mkdir -p ~/data/plastid/Atha_1001/ena
cd ~/data/plastid/Atha_1001/ena

cat SraRunTable.csv |
    mlr --icsv --otsv cat \
    > SraRunTable.tsv

cat SraRunTable.tsv |
    tsv-filter -H \
        --str-in-fld tissue:leaf \
        --ge AvgSpotLen:90

curl -o accessions.csv \
    https://tools.1001genomes.org/api/accessions.csv?query=SELECT%20*%20FROM%20tg_accessions%20ORDER%20BY%20id

echo -e 'Ecotype\tName\tCS Number\tCountry\tLat\tLong\tCollector\tAdmixture Group\tSequenced by' \
    > accessions.tsv

cat accessions.csv |
    mlr --icsv --implicit-csv-header --otsv cat |
    sed 1d |
    tsv-select -f 1,3,10,4,6,7,8,11,2 \
    >> accessions.tsv

cat accessions.tsv |
    tsv-join -H --key-fields "Ecotype" \
        -f SraRunTable.tsv \
        --append-fields tissue,AvgSpotLen,Bases,Experiment |
    tsv-filter -H \
        --str-in-fld tissue:leaf \
        --ge Bases:1000000000 \
        --le Bases:10000000000 \
        --ge AvgSpotLen:90 \
        --regex Name:'^[\w\d-]+$' |
    tsv-select -H -f Experiment,Name,Country |
    mlr --itsv --ocsv cat |
    sed 1d \
    > source.csv

anchr ena info | perl - -v source.csv > ena_info.yml
anchr ena prep | perl - ena_info.yml --ascp

rgr md ena_info.tsv --fmt | head -n 20

# sort by SRR
cat ena_info.ascp.sh |
    sort > tmp.sh &&
    mv tmp.sh ena_info.ascp.sh

cat ena_info.ascp.sh |
    parallel --no-run-if-empty -j 2 "{}"

# Skips
cat ena_info.csv |
    mlr --icsv --otsv cat |
    tsv-select -H -f name,srr,bases |
    parallel --colsep '\t' --no-run-if-empty --linebuffer -k -j 4 '
        if [ ! -f {2}_1.fastq.gz ]; then
            echo 1>&2 {1}
            echo {1}
            exit;
        fi
        if [ ! -f {2}_2.fastq.gz ]; then
            echo 1>&2 {1}
            echo {1}
            exit;
        fi

        LENGTH=$(gzip -dcf {2}_1.fastq.gz |
            head -n 100 |
            faops n50 -H stdin
        )
        [[ $LENGTH -le 90 ]] && ( echo 1>&2 {1}; echo {1}; exit; )

        LENGTH=$(gzip -dcf {2}_2.fastq.gz |
            head -n 100 |
            faops n50 -H stdin
        )
        [[ $LENGTH -le 90 ]] && ( echo 1>&2 {1}; echo {1}; exit;  )
    ' |
    sort -r |
    uniq \
    > skip.lst

```

| name        | srx       | platform | layout | ilength | srr        |      spots | bases |
|-------------|-----------|----------|--------|---------|------------|-----------:|-------|
| 11C1        | SRX972788 | ILLUMINA | PAIRED |         | SRR1946105 | 20,357,230 | 3.79G |
| 328PNA062   | SRX972701 | ILLUMINA | PAIRED |         | SRR1946018 | 73,235,208 | 6.62G |
| 627ME-13Y1  | SRX972634 | ILLUMINA | PAIRED |         | SRR1945951 | 44,902,951 | 8.28G |
| 627ME-1MI1  | SRX972635 | ILLUMINA | PAIRED |         | SRR1945952 | 48,675,298 | 8.98G |
| 627RMX-1MN4 | SRX972632 | ILLUMINA | PAIRED |         | SRR1945949 | 37,102,228 | 6.82G |
| 627RMX-1MN5 | SRX972633 | ILLUMINA | PAIRED |         | SRR1945950 | 44,085,175 | 8.11G |
| ANH-1       | SRX972422 | ILLUMINA | PAIRED |         | SRR1945739 | 10,783,643 | 2.03G |
| ARGE-1-15   | SRX973146 | ILLUMINA | PAIRED |         | SRR1946463 | 25,951,374 | 4.83G |
| ARR-17      | SRX973157 | ILLUMINA | PAIRED |         | SRR1946474 | 14,893,935 | 2.77G |
| Aa-0        | SRX972490 | ILLUMINA | PAIRED |         | SRR1945807 |  9,051,264 | 1.7G  |
| Abd-0       | SRX972484 | ILLUMINA | PAIRED |         | SRR1945801 | 13,961,306 | 2.63G |
| Adam-1      | SRX972881 | ILLUMINA | PAIRED |         | SRR1946198 | 12,010,826 | 2.24G |
| Ag-0        | SRX972432 | ILLUMINA | PAIRED |         | SRR1945749 | 10,141,217 | 1.91G |
| Aiell-1     | SRX972915 | ILLUMINA | PAIRED |         | SRR1946232 | 12,869,053 | 2.4G  |
| Aitba-1     | SRX972878 | ILLUMINA | PAIRED |         | SRR1946195 | 11,730,031 | 2.18G |
| Ak-1        | SRX972485 | ILLUMINA | PAIRED |         | SRR1945802 |  8,524,299 | 1.6G  |
| Alst-1      | SRX972486 | ILLUMINA | PAIRED |         | SRR1945803 |  9,347,634 | 1.76G |
| Alt-1       | SRX973020 | ILLUMINA | PAIRED |         | SRR1946337 | 24,312,469 | 4.53G |

## Symlink

* 采用的倍数因子值: `2`

* Rsync to hpcc

```shell
rsync -avP \
    ~/data/plastid/Atha_1001/ \
    wangq@202.119.37.251:data/plastid/Atha_1001

# rsync -avP wangq@202.119.37.251:data/plastid/Atha_1001/ ~/data/plastid/Atha_1001

```

```shell
cd ~/data/plastid/Atha_1001/

export FOLD=2
export GENOME_SIZE=$(
    cat ~/data/plastid/genome/col_0/chr.sizes |
        tsv-summarize --sum 2
)

cat ena/ena_info.csv |
    mlr --icsv --otsv cat |
    tsv-select -H -f name,srr,bases |
    tsv-join -H --filter-file ena/skip.lst --key-fields name --exclude |
    perl -nla -F'\t' -e '
        BEGIN { our %seen }
        /^name/ and next;
        $seen{$F[0]} and next;
        my $bases = $F[2];
        $bases =~ s/G$//;
        my $cutoff = $bases * 1000 * 1000 * 1000 / $ENV{GENOME_SIZE} * $ENV{FOLD};
        $cutoff = int $cutoff;
        print join qq(\t), ($F[0], $F[1], $cutoff);
        $seen{$F[0]}++;
    ' \
    > opts.tsv

wc -l opts.tsv
# 705

cat opts.tsv |
    parallel --colsep '\t' --no-run-if-empty --linebuffer -k -j 1 '
        if [ ! -f ena/{2}_1.fastq.gz ]; then
            exit;
        fi
        if [ ! -f ena/{2}_2.fastq.gz ]; then
            exit;
        fi

        if [ -f {1}.tar.gz ]; then
            exit;
        fi

        mkdir -p {1}/1_genome
        pushd {1}/1_genome

        cp ../../genome/genome.fa genome.fa
        popd > /dev/null

        mkdir -p {1}/2_illumina
        pushd {1}/2_illumina

        ln -fs ../../ena/{2}_1.fastq.gz R1.fq.gz
        ln -fs ../../ena/{2}_2.fastq.gz R2.fq.gz
        popd > /dev/null
    '

```

* Rsync non-processed files to hpcc

```shell
cd ~/data/plastid/Atha_1001/

cat opts.tsv |
    parallel --colsep '\t' --no-run-if-empty --linebuffer -k -j 1 '
        if [ -f {1}.tar.gz ]; then
            exit;
        fi

        find ena -type f -name "*{2}*"
    ' \
    > rsync.lst

rsync -avP \
    --files-from=rsync.lst \
    ~/data/plastid/Atha_1001/ \
    wangq@202.119.37.251:data/plastid/Atha_1001

```

## Run

```shell
cd ~/data/plastid/Atha_1001/

cat opts.tsv | #head -n 400 | #tail -n 10 |
    parallel --colsep '\t' --no-run-if-empty --linebuffer -k -j 1 '
        if [ -f {1}.tar.gz ]; then
            exit;
        fi

        if [ ! -d {1} ]; then
            exit;
        fi

        if [ ! -e {1}/2_illumina/R1.fq.gz ]; then
            exit;
        fi
        if [ ! -e {1}/2_illumina/R2.fq.gz ]; then
            exit;
        fi

        if bjobs -w | tr -s " " | cut -d " " -f 7 | grep -w "^{1}$"; then
            echo Job {1} exists
            exit;
        fi

        cd {1}

        echo {1}

        rm *.sh
        anchr template \
            --genome 1000000 \
            --parallel 24 \
            --xmx 80g \
            \
            --fastqc \
            --insertsize \
            --kat \
            \
            --trim "--dedupe --cutoff {3} --cutk 31" \
            --qual "25" \
            --len "60" \
            --filter "adapter artifact" \
            \
            --bwa Q25L60 \
            --gatk

        bsub -q mpi -n 24 -J "{1}" "
            bash 2_fastqc.sh
            bash 2_insert_size.sh
            bash 2_kat.sh
            bash 2_trim.sh
            bash 9_stat_reads.sh
            bash 3_bwa.sh
            bash 3_gatk.sh
        "
    '

```

## Pack and clean

```shell
cd ~/data/plastid/Atha_1001/

cat opts.tsv |
    parallel --colsep '\t' --no-run-if-empty --linebuffer -k -j 4 '
        if [ -f {1}.tar.gz ]; then
            echo "==> {1} .tar.gz"
            exit;
        fi

        if [ ! -f {1}/3_gatk/R.filtered.vcf ]; then
            echo "==> {1} 3_gatk"
            exit;
        fi

#        if [ ! -f {1}/7_merge_anchors/anchor.merge.fasta ]; then
#            echo "==> {1} 7_merge_anchors"
#            exit;
#        fi
#
#        if [ ! -d {1}/9_quast ]; then
#            echo "==> {1} 9_quast"
#            exit;
#        fi

        echo "==> Clean {1}"
        bash {1}/0_cleanup.sh

        echo "==> Create {1}.tar.gz"

        tar -czvf {1}.tar.gz \
            {1}/1_genome/genome.fa \
            {1}/2_illumina/fastqc \
            {1}/2_illumina/insert_size \
            {1}/2_illumina/kat \
            {1}/3_bwa/join.tsv \
            {1}/3_bwa/R.dedup.metrics \
            {1}/3_bwa/R.wgs.metrics \
            {1}/3_bwa/R.sort.bam \
            {1}/3_bwa/R.sort.bai \
            {1}/3_gatk \
            {1}/7_merge_anchors/anchor.merge.fasta \
            {1}/7_merge_anchors/others.non-contained.fasta \
            {1}/8_megahit/anchor/anchor.fasta \
            {1}/8_megahit/megahit.non-contained.fasta \
            {1}/8_mr_megahit/anchor/anchor.fasta \
            {1}/8_mr_megahit/megahit.non-contained.fasta \
            {1}/8_spades/anchor/anchor.fasta \
            {1}/8_spades/spades.non-contained.fasta \
            {1}/8_mr_spades/anchor/anchor.fasta \
            {1}/8_mr_spades/spades.non-contained.fasta \
            {1}/8_platanus/anchor/anchor.fasta \
            {1}/8_platanus/platanus.non-contained.fasta \
            {1}/9_quast \
            {1}/*.md

        echo
    '

cat opts.tsv |
    parallel --colsep '\t' --no-run-if-empty --linebuffer -k -j 1 '
        if [ -f {1}.tar.gz ]; then
            if [ -d {1} ]; then
                echo "==> Remove {1}/"
                rm -fr {1}
            fi
        fi
    '

```

* Remove processed files

```shell
cd ~/data/plastid/Atha_1001/

cat opts.tsv |
    parallel --colsep '\t' --no-run-if-empty --linebuffer -k -j 1 '
        if [ ! -f {1}.tar.gz ]; then
            exit;
        fi

        find ena -type f -name "*{2}*"
    ' |
    xargs rm

```

* Unpack

```shell
cd ~/data/plastid/Atha_1001/

cat opts.tsv |
    parallel --colsep '\t' --no-run-if-empty --linebuffer -k -j 1 '
        if [ -d {1} ]; then
            echo "==> {1} exists"
            exit;
        fi

        if [ ! -f {1}.tar.gz ]; then
            echo "==> {1}.tar.gz not exists"
            exit;
        fi

        tar -xzvf {1}.tar.gz
        rm {1}.tar.gz
    '

```

## VCF

```shell
cd ~/data/plastid/Atha_1001/

mkdir -p vcf

cat opts.tsv |
    parallel --colsep '\t' --no-run-if-empty --linebuffer -k -j 4 '
        if [ ! -f {1}.tar.gz ]; then
            echo "==> {1}.tar.gz not exists"
            exit;
        fi

        tar -xOzvf {1}.tar.gz {1}/3_gatk/R.filtered.vcf |
            bcftools reheader --samples <(echo {1}) |
            bcftools view \
                --apply-filters PASS --types snps --max-alleles 2 --targets Pt -Oz |
            bcftools view --include "AF>0.01" -Oz -o vcf/{1}.vcf.gz

        bcftools index -f vcf/{1}.vcf.gz
    '

bcftools merge --merge all -l <(
        cat opts.tsv |
            cut -f 1 |
            parallel -k -j 1 ' [ -f vcf/{}.vcf.gz ] && echo "vcf/{}.vcf.gz" '
    ) \
    > Atha_1001.vcf

rm -fr vcf

```

